File: ProductionConfigTool/packet_builder.py
import time
import datetime

class PacketBuilder:
    """
    Builds a 40-byte packet according to the protocol specification
    for sending configuration settings to production line devices
    """
    
    def __init__(self, config_data):
        """
        Initialize packet builder with configuration data
        
        Args:
            config_data (dict): Dictionary containing configuration parameters
        """
        self.config_data = config_data
    
    def build_packet(self):
        """
        Build the 40-byte packet according to the protocol specification
        
        Returns:
            bytes: 40-byte packet
        """
        # Initialize 40-byte array with zeros
        packet = bytearray(40)
        
        # Fill in packet data according to protocol
        
        # [0] STX
        packet[0] = 0xDA
        
        # [1] Product type
        packet[1] = self.config_data['product_type']
        
        # [2]~[3] ID (MAC address lower 2 bytes)
        mac_bytes = bytes.fromhex(self.config_data['mac_address'])
        packet[2] = mac_bytes[0] if len(mac_bytes) > 0 else 0
        packet[3] = mac_bytes[1] if len(mac_bytes) > 1 else 0
        
        # [4] Data length is bytes [5]~[36] (32 bytes)
        packet[4] = 32
        
        # [5] Circuit configuration
        packet[5] = (
            (self.config_data['light_circuits'] & 0x0F) |        # bits 0-3: light circuits
            ((self.config_data['outlet_circuits'] & 0x03) << 4) | # bits 4-5: outlet circuits
            ((self.config_data['dimming_type'] & 0x03) << 6)      # bits 6-7: dimming type
        )
        
        # [6] Delay time
        packet[6] = self.config_data['delay_time'] & 0xFF
        
        # [7] SUB ID
        packet[7] = self.config_data['sub_id'] & 0xFF
        
        # [8] IR presence
        packet[8] = self.config_data['ir_present'] & 0x01
        
        # [9] Scenario
        packet[9] = self.config_data['scenario'] & 0x07
        
        # [10] Communication company
        packet[10] = self.config_data['comm_company'] & 0x07
        
        # [11] 3Way
        packet[11] = self.config_data['three_way'] & 0x01
        
        # [12] Overload protection and emergency call
        packet[12] = (
            (self.config_data['overload_protection'] & 0x0F) |   # bits 0-3: overload protection
            ((self.config_data['emergency_call'] & 0x0F) << 4)   # bits 4-7: emergency call
        )
        
        # [13]~[16] Outlet 1 values
        outlet1_learn = self.config_data['outlet1_learn_value']
        outlet1_current = self.config_data['outlet1_current_value']
        
        packet[13] = outlet1_learn & 0xFF         # Low byte
        packet[14] = (outlet1_learn >> 8) & 0xFF  # High byte
        packet[15] = outlet1_current & 0xFF       # Low byte
        packet[16] = (outlet1_current >> 8) & 0xFF # High byte
        
        # [17]~[20] Outlet 2 values
        outlet2_learn = self.config_data['outlet2_learn_value']
        outlet2_current = self.config_data['outlet2_current_value']
        
        packet[17] = outlet2_learn & 0xFF         # Low byte
        packet[18] = (outlet2_learn >> 8) & 0xFF  # High byte
        packet[19] = outlet2_current & 0xFF       # Low byte
        packet[20] = (outlet2_current >> 8) & 0xFF # High byte
        
        # [21] Relay ON/OFF status
        packet[21] = self.config_data['relay_status'] & 0xFF
        
        # [22] Outlet 1 mode
        packet[22] = self.config_data['outlet1_mode'] & 0x01
        
        # [23] Outlet 2 mode
        packet[23] = self.config_data['outlet2_mode'] & 0x01
        
        # [24] Reserved (set to 0)
        packet[24] = 0
        
        # [25] Sleep mode
        packet[25] = self.config_data['sleep_mode'] & 0x01
        
        # [26] Delay mode
        packet[26] = self.config_data['delay_mode'] & 0x01
        
        # [27] Dimming current value
        packet[27] = self.config_data['dimming_value'] & 0xFF
        
        # [28] Color temperature current value
        packet[28] = self.config_data['color_temp_value'] & 0xFF
        
        # [29]~[32] Current date/time from Windows
        now = datetime.datetime.now()
        year_offset = now.year - 2020
        
        packet[29] = ((year_offset & 0x0F) << 4) | (now.month & 0x0F)  # Year (offset from 2020) and month
        packet[30] = now.day & 0xFF    # Day
        packet[31] = now.hour & 0xFF   # Hour
        packet[32] = now.minute & 0xFF # Minute
        
        # [33]~[34] Reserved (set to 0)
        packet[33] = 0
        packet[34] = 0
        
        # [35]~[36] Version information
        packet[35] = 0x03  # Version high byte
        packet[36] = 0x13  # Version low byte
        
        # [37] XOR checksum of bytes [6]~[36]
        xor_checksum = 0
        for i in range(6, 37):
            xor_checksum ^= packet[i]
        packet[37] = xor_checksum
        
        # [38] ADD checksum of bytes [6]~[36]
        add_checksum = 0
        for i in range(6, 37):
            add_checksum += packet[i]
        packet[38] = add_checksum & 0xFF  # Take only lower byte
        
        # [39] ETX
        packet[39] = 0x25
        
        return bytes(packet)
    
    def validate_packet(self, packet):
        """
        Validate packet checksums and structure
        
        Args:
            packet (bytes): Packet to validate
            
        Returns:
            bool: True if valid, False if invalid
        """
        if len(packet) != 40:
            return False
            
        # Check STX and ETX
        if packet[0] != 0xDA or packet[39] != 0x25:
            return False
            
        # Calculate and verify XOR checksum
        xor_checksum = 0
        for i in range(6, 37):
            xor_checksum ^= packet[i]
            
        if xor_checksum != packet[37]:
            return False
            
        # Calculate and verify ADD checksum
        add_checksum = 0
        for i in range(6, 37):
            add_checksum += packet[i]
            
        if (add_checksum & 0xFF) != packet[38]:
            return False
            
        return True

File: ProductionConfigTool/PacketBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace ProductionTesterApp
{
    public class PacketBuilder
    {
        private Dictionary<string, object> configData;
        private const byte STX = 0x02;  // 시작 문자
        private const byte ETX = 0x03;  // 종료 문자

        /// <summary>
        /// 패킷 빌더 초기화
        /// </summary>
        /// <param name="configData">설정 데이터 딕셔너리</param>
        public PacketBuilder(Dictionary<string, object> configData)
        {
            this.configData = configData;
        }

        /// <summary>
        /// 프로토콜 사양에 따라 40바이트 패킷 구성
        /// </summary>
        /// <returns>40바이트 패킷</returns>
        public byte[] BuildPacket()
        {
            byte[] packet = new byte[40];  // 40바이트 패킷

            // 1. STX (시작 문자)
            packet[0] = STX;

            // 2. 제품 유형
            packet[1] = Convert.ToByte(configData["product_type"]);

            // 3. MAC 주소 (마지막 2바이트)
            string macStr = configData["mac_address"].ToString();
            if (macStr.Length == 4)
            {
                packet[2] = Convert.ToByte(macStr.Substring(0, 2), 16);
                packet[3] = Convert.ToByte(macStr.Substring(2, 2), 16);
            }
            else
            {
                packet[2] = 0x00;
                packet[3] = 0x00;
            }

            // 4. 데이터 길이
            packet[4] = 33;  // STX, 제품유형, MAC 주소 2바이트, 데이터 길이, XOR, ADD, ETX를 제외한 길이

            // 5. 조명 회로 설정
            packet[5] = Convert.ToByte(configData["light_circuits"]);

            // 6. 콘센트 회로 설정
            packet[6] = Convert.ToByte(configData["outlet_circuits"]);

            // 7. 디밍 유형 설정
            packet[7] = Convert.ToByte(configData["dimming_type"]);

            // 8. 기타 설정 필드 (기본값으로 채우기)
            for (int i = 8; i < 36; i++)
            {
                packet[i] = 0x00;
            }

            // 9. 현재 날짜 및 시간
            DateTime now = DateTime.Now;
            packet[29] = (byte)(now.Year - 2020);  // 2020년 기준 오프셋
            packet[30] = (byte)now.Month;
            packet[31] = (byte)now.Day;
            packet[32] = (byte)now.Hour;
            packet[33] = (byte)now.Minute;

            // 10. XOR 체크섬
            packet[36] = CalculateChecksumXOR(packet, 1, 36);

            // 11. ADD 체크섬
            packet[37] = CalculateChecksumADD(packet, 1, 36);

            // 12. ETX (종료 문자)
            packet[38] = ETX;

            // 13. 더미 바이트 (0x00)
            packet[39] = 0x00;

            return packet;
        }

        /// <summary>
        /// 패킷 체크섬 및 구조 유효성 검사
        /// </summary>
        /// <param name="packet">검사할 패킷</param>
        /// <returns>유효한 경우 true, 그렇지 않은 경우 false</returns>
        public bool ValidatePacket(byte[] packet)
        {
            if (packet == null || packet.Length != 40)
                return false;

            // STX 및 ETX 확인
            if (packet[0] != STX || packet[38] != ETX)
                return false;

            // 체크섬 확인
            byte xorChecksum = CalculateChecksumXOR(packet, 1, 36);
            byte addChecksum = CalculateChecksumADD(packet, 1, 36);

            return packet[36] == xorChecksum && packet[37] == addChecksum;
        }

        /// <summary>
        /// XOR 체크섬 계산
        /// </summary>
        /// <param name="data">체크섬을 계산할 데이터</param>
        /// <param name="start">시작 인덱스</param>
        /// <param name="end">종료 인덱스 (포함)</param>
        /// <returns>XOR 체크섬</returns>
        private byte CalculateChecksumXOR(byte[] data, int start, int end)
        {
            byte result = 0;
            for (int i = start; i <= end; i++)
            {
                result ^= data[i];
            }
            return result;
        }

        /// <summary>
        /// ADD 체크섬 계산 (모든 바이트의 합, 하위 바이트만 유지)
        /// </summary>
        /// <param name="data">체크섬을 계산할 데이터</param>
        /// <param name="start">시작 인덱스</param>
        /// <param name="end">종료 인덱스 (포함)</param>
        /// <returns>ADD 체크섬 (하위 바이트만)</returns>
        private byte CalculateChecksumADD(byte[] data, int start, int end)
        {
            int sum = 0;
            for (int i = start; i <= end; i++)
            {
                sum += data[i];
            }
            return (byte)(sum & 0xFF);  // 하위 바이트만 유지
        }
    }
}